lib/matplotlib/tests/test_violinplot_datetime.py

```diff
@@ -0,0 +1,95 @@
+"""
+Unit tests for proactive validation of datetime
+positions and timedelta widths in violinplot.
+"""
+
+import datetime
+import pytest
+
+import matplotlib.pyplot as plt
+
+
+def make_vpstats():
+    """Create minimal valid stats for a violin plot."""
+
+
+def violin_plot_stats():
+    # Stats for violin plot
+    datetimes = [
+        datetime.datetime(2023, 2, 10),
+        datetime.datetime(2023, 5, 18),
+        datetime.datetime(2023, 6, 6)
+    ]
+    return [{
+        'coords': datetimes,
+        'vals': [0.1, 0.5, 0.2],
+        'mean': datetimes[1],
+        'median': datetimes[1],
+        'min': datetimes[0],
+        'max': datetimes[-1],
+        'quantiles': datetimes
+    }, {
+        'coords': datetimes,
+        'vals': [0.2, 0.3, 0.4],
+        'mean': datetimes[2],
+        'median': datetimes[2],
+        'min': datetimes[0],
+        'max': datetimes[-1],
+        'quantiles': datetimes
+    }]
+
+
+def test_datetime_positions_with_float_widths_raises():
+    """Test that datetime positions with float widths raise TypeError."""
+    fig, ax = plt.subplots()
+    try:
+        vpstats = violin_plot_stats()
+        positions = [datetime.datetime(2020, 1, 1), datetime.datetime(2021, 1, 1)]
+        widths = [0.5, 1.0]
+        with pytest.raises(TypeError,
+        match="positions are datetime/date.*widths as datetime\\.timedelta"):
+            ax.violin(vpstats, positions=positions, widths=widths)
+    finally:
+        plt.close(fig)
```

Comment:
The test infrastructure takes care of closing figures. You don't need to do that yourself.
```suggestion
    vpstats = violin_plot_stats()
    positions = [datetime.datetime(2020, 1, 1), datetime.datetime(2021, 1, 1)]
    widths = [0.5, 1.0]
    with pytest.raises(TypeError,
            match="positions are datetime/date.*widths as datetime\\.timedelta"):
        ax.violin(vpstats, positions=positions, widths=widths)
```
Holds also for the other tests.

You could also inline `violin_plot_stats() as `ax.violin(vpstats(), ...)` which has less indirection.